File: files/2026-06-07/CSEC 07062026.R
library(tidyverse)
library(readxl)
library(igraph)

path <- "2026-06-07/Challenge 134.xlsx"
input <- read_excel(path, range = "B2:C14")
test <- read_excel(path, range = "E2:G14") %>%
  replace_na(list(Descendants = ""))

edges <- input %>%
  filter(!is.na(`Depends On`), `Depends On` != "") %>%
  transmute(from = `Depends On`, to = `Task ID`)
g <- graph_from_data_frame(edges, vertices = input$`Task ID`)
result <- tibble(Task = V(g)$name) %>%
  mutate(
    downstream = map(
      Task,
      ~ subcomponent(g, .x, mode = "out")$name |> setdiff(.x)
    ),
    Count = map_int(downstream, length),
    Descendants = map_chr(downstream, ~ str_c(.x, collapse = ", "))
  ) %>%
  select(Task, Count, Descendants)

all.equal(result, test)

File: files/2026-06-07/CSEC 07062026.py
import pandas as pd
import networkx as nx

path = "2026-06-07/Challenge 134.xlsx"
input = pd.read_excel(path, usecols="B:C", nrows=13, skiprows=1)
test = pd.read_excel(path, usecols="E:G", nrows=13, skiprows=1).fillna("")

edges = (
    input.dropna(subset=["Depends On"])
    .assign(**{"Depends On": lambda x: x["Depends On"].str.split(",")})
    .explode("Depends On")
    .assign(**{"Depends On": lambda x: x["Depends On"].str.strip()})
    .query("`Depends On` != ''")
)
G = nx.DiGraph()
G.add_nodes_from(input["Task ID"])
G.add_edges_from(zip(edges["Depends On"], edges["Task ID"]))
result = pd.DataFrame({"Task": input["Task ID"]})
result["Downstream"] = result["Task"].apply(lambda x: sorted(nx.descendants(G, x)))
result["Count"] = result["Downstream"].str.len()
result["Descendants"] = result["Downstream"].apply(", ".join)
result = result[["Task", "Count", "Descendants"]]
print(result.equals(test))
# True
